statemachine/src/main/java/com/amazonaws/stepfunctions/cloudformation/statemachine/DefinitionProcessor.java

```diff
@@ -79,20 +79,26 @@ private static int getNumDefinitionsInModel(final ResourceModel model) {
      * @param metricsRecorder The MetricsRecorder object used for collecting anonymous property usage metrics
      */
     public static void processDefinition(final AmazonWebServicesClientProxy proxy, final ResourceModel model, final MetricsRecorder metricsRecorder) {
-        if (model.getDefinitionS3Location() != null) {
-            model.setDefinitionString(fetchS3Definition(model.getDefinitionS3Location(), proxy, metricsRecorder));
+        String definition = "";
+        if(model.getDefinitionString() != null){
+            definition = model.getDefinitionString();
         }
-
-        if (model.getDefinition() != null) {
-            model.setDefinitionString(convertDefinitionObjectToString(model.getDefinition()));
+        else{
+            definition = model.getDefinitionS3Location() != null ? fetchS3Definition(model.getDefinitionS3Location(), proxy) : convertDefinitionObjectToString(model.getDefinition());
         }
 
         if (model.getDefinitionSubstitutions() != null) {
-            model.setDefinitionString(transformDefinition(model.getDefinitionString(), model.getDefinitionSubstitutions()));
+            definition = transformDefinition(definition, model.getDefinitionSubstitutions());
+        }
+
+        if(model.getDefinitionS3Location() != null){
```

Comment:
Question: Do YAML substitutions break if we do the substitutions before we convert the definition to a JSON string?